### PR TITLE
feat: local display names for nodes

### PR DESCRIPTION
**Pull request by gilindoesgit**
# Local display names for nodes

Users can set a **local display name** for any node. That name is shown in the app instead of the device’s long/short name, and is stored only on the device (not sent over the mesh).

---

## Summary

<table>
<tr><td><strong>Storage</strong></td><td><code>NodeDisplayNameStore</code> (UserDefaults), keyed by <strong>node number</strong></td></tr>
<tr><td><strong>Scope</strong></td><td>Node list, node detail, messages (user list &amp; relay), compass waypoint name</td></tr>
<tr><td><strong>Edit entry points</strong></td><td>Long-press node → “Set display name”; Node detail → “Display name” row</td></tr>
</table>

---

## UI

<ul>
<li><strong>Node list</strong> – Rows show the custom display name when set, otherwise the device long/short name.</li>
<li><strong>Long-press menu</strong> – “Set display name” opens a sheet to set or clear the name.</li>
<li><strong>Node detail</strong> – Navigation title and a “Display name” row; tapping the row opens the edit sheet.</li>
<li><strong>Edit sheet</strong> – Text field for the local name, with options to save or clear the display name.</li>
<li><strong>Messages</strong> – User list and relay text use <code>displayLongName</code> / <code>displayShortName</code>.</li>
</ul>

---

## Technical details

- **Key:** Display names are stored and looked up by <strong>node number</strong> (<code>num</code>), the unique node ID.
- **Persistence:** UserDefaults key <code>nodeDisplayNames</code>; JSON dictionary <code>[String: String]</code> with node num as string key.
- **Model:** <code>UserEntity</code> extensions <code>displayLongName</code> and <code>displayShortName</code>; they return the stored name if set, otherwise the existing <code>longName</code> / <code>shortName</code>.
- **UI refresh:** <code>NodeDisplayNameStore.didChangeNotification</code> is posted when a name is set or cleared; detail view subscribes to refresh the title.

---

## Checklist

- [x] Display name keyed by node number only
- [x] Shown in list, detail, and messages; editable from list context menu and detail row
- [x] Clear display name restores device name
- [x] No change to device identity or protocol; local-only

## Summary by CodeRabbit

* **New Features**
  * Added local node display names you can set from the node list (long-press/context menu) and node details, and clear them when needed.
  * Display names now show across the app UI (node list/detail and messages/conversation titles), without affecting the node’s real identity used for QR/contact sharing.
* **Bug Fixes**
  * Improved consistency so relay titles/waypoints and accessibility/VoiceOver naming reliably prefer display names.
* **Documentation**
  * Updated the Nodes help and What’s New content to explain display names are device-local only and how to use them.

### CODE REVIEW

**Comment by garthvh** [comment]
This feature was in the app before and just caused confusion, without this data persisting on the node it just seems like a way to frustrate users.

**Comment by TMTwatch** [comment]
Maybe a notes field could be added instead? A small icon n next in the node list could signal that there's added info in the notes

**Review by garthvh — DISMISSED** [review]
Thanks for the PR — local display names is a genuinely useful feature and the core idea is well thought out. `NodeDisplayNameStore` is clean, the notification-based refresh pattern is correct, and the `displayLongName`/`displayShortName` computed properties are a nice drop-in replacement. However there are several blocking issues that need to be addressed before this can merge.

---

## Critical — Wrong extension files / Core Data assumptions

This PR was authored against an old version of the repo. **Core Data has been fully removed from this project** — all persistence is SwiftData. The affected files have moved:

| PR targets | Correct path on `main` |
|---|---|
| `Meshtastic/Extensions/CoreData/UserEntityExtension.swift` | `Meshtastic/Extensions/SwiftData/UserEntityExtension.swift` |
| `Meshtastic/Extensions/CoreData/NodeInfoEntityExtension.swift` | `Meshtastic/Extensions/SwiftData/NodeInfoEntityExtension.swift` |
| `Meshtastic/Extensions/CoreData/MessageEntityExtension.swift` | `Meshtastic/Extensions/SwiftData/MessageEntityExtension.swift` |

Please rebase onto current `main` and redirect all changes to the SwiftData extension files.

## Critical — NSManagedObjectID / Core Data API in NodeInfoEntityExtension

The diff adds:

```swift
extension NodeInfoEntity: Identifiable {
    public var id: NSManagedObjectID { objectID }
}
```

`NSManagedObjectID` and `objectID` are Core Data APIs. `NodeInfoEntity` is a SwiftData `@Model` — it already conforms to `Identifiable` through SwiftData via `persistentModelID`. This block needs to be removed entirely.

## Critical — `import CoreData` in EditNodeDisplayNameView.swift

`EditNodeDisplayNameView` imports CoreData but does not use it. Remove the import — the view only needs `SwiftUI`.

## Critical — NodeList uses @Query, not NSFetchRequest

The PR's `FilteredNodeList` initialiser sets up an `NSFetchRequest<NodeInfoEntity>` — but on `main`, `NodeList` uses `@Environment(\.modelContext)` and `@Query` (SwiftData). These APIs are incompatible. After rebasing, the node-for-display-name binding needs to be wired into the existing SwiftData query-driven list.

## Warning — Force-unwrap in MessageEntityExtension.swift

```swift
let name = users.first!.displayLongName  // crash if users is somehow empty
```

The original code safely used optional binding. Please restore a guard or use optional chaining instead.

## Warning — Indentation regressions

The diff introduces broken indentation in `NodeDetail.swift`, `NodeListItem.swift`, `NodeRow.swift`, and `NodeList.swift` — mixed tabs/spaces and misaligned braces. The `.sheet` call in `NodeDetail` is de-indented out of its `ScrollViewReader` scope. Please run SwiftLint before submitting (`scripts/lint`).

## Warning — docs/PULL_REQUEST_DISPLAY_NAMES.md should be removed

This file is a copy of the PR description dropped into `docs/`. It is not a user guide or developer doc. Please delete it. If you would like to document the feature, add a section to `docs/user/nodes.md` instead (per the view-to-doc mapping in CONTRIBUTING.md).

## Warning — No tests

`NodeDisplayNameStore` is the most independently testable piece here and has no coverage. Please add a Swift Testing suite (`import Testing`, `@Suite`, `@Test`, `#expect`) covering at minimum: set/get, clear, persistence across instances, and notification posting. See `MeshtasticTests/RouterTests.swift` for the project's testing conventions.

---

## What to keep as-is

- `NodeDisplayNameStore.swift` — solid implementation, keep it
- `EditNodeDisplayNameView.swift` — clean UI, keep it (just remove `import CoreData`)
- `displayLongName` / `displayShortName` computed properties on `UserEntity` — correct approach, just apply them to the right SwiftData extension file
- The `NodeDisplayNameStore.didChangeNotification` + `.id(displayNameRefresh)` refresh pattern in `NodeDetail` — correct, keep it

**Comment by garthvh** [comment]
Hi @gilindoesgit — thanks for this! To get it mergeable I had to re-implement it against the app's current architecture: the branch predates the CoreData → SwiftData migration (#1752, #1886), and it edited `Meshtastic/Extensions/CoreData/*.swift` files that no longer exist (moved/renamed to `Extensions/SwiftData/*`), plus `NodeList.swift`/`NodeDetail.swift` changed shape since (Status Message row, deleted-model guards, etc.). Rather than a textual conflict resolution, I force-pushed a from-scratch re-port that keeps the same design — `NodeDisplayNameStore` (UserDefaults JSON), `displayLongName`/`displayShortName` on `UserEntity`, the same edit sheet and surfaces — just wired into today's files. Your original commits are still visible in this PR's timeline if you want to compare.

A couple of small decisions I made along the way, flagging in case you'd rather go a different way:
- Left the CoreData `NodeInfoEntity: Identifiable` shim out — SwiftData's `@Model` already synthesizes that.
- Didn't extend this to `NodeListItemCompact.swift` (the compact node-list density mode) since it postdates your original PR — worth a quick follow-up for consistency if you're up for it.
- Swapped the ad-hoc `docs/PULL_REQUEST_DISPLAY_NAMES.md` for a short section in `docs/user/nodes.md`, matching how this repo documents user-facing features (it's bundled into the in-app docs).

Build is clean and the existing navigation/router/snapshot test suites pass. Let me know if you'd like anything adjusted.

File: Meshtastic/Resources/docs/markdown/user/nodes.md [hunk 1]
```diff
@@ -96,12 +96,19 @@ The full node row shows the circle avatar, battery level, encryption status, las
 Long-press any node in the list to access quick actions:
 
 - **Add to favorites / Remove from favorites** — star important nodes so they appear at the top of the list
+- **Set display name** — give a node a local nickname (see [Display Names](#display-names) below)
```
[the remark this answers is not in the document]
Comment (garthvh) [reply Meshtastic/Resources/docs/markdown/user/nodes.md]: Fixed in 0abede4d — replaced the broken `#display-names` anchor with plain text since the docs pipeline does not emit heading ids.

File: Meshtastic/Views/Nodes/Helpers/EditNodeDisplayNameView.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+//  Sheet to set or clear a local display name for a node.
+//
+
+import SwiftUI
+
+struct EditNodeDisplayNameView: View {
+	@Environment(\.dismiss) private var dismiss
+	let node: NodeInfoEntity
+	@State private var displayName: String = ""
+	@State private var hasChanges: Bool = false
```
[the remark this answers is not in the document]
Comment (garthvh) [reply Meshtastic/Views/Nodes/Helpers/EditNodeDisplayNameView.swift]: Fixed in 0abede4d — now tracks `initialDisplayName` from `onAppear` and compares against it, so Save is only enabled on a real edit.

File: Meshtastic/Views/Nodes/Helpers/NodeDetail.swift [hunk 3]
```diff
@@ -70,8 +75,14 @@ struct NodeDetail: View {
...
 					.sheet(isPresented: $showingCompassSheet) {
-						CompassView(waypointLocation: latestPosition?.nodeCoordinate ?? nil, waypointLongName: node.user?.longName ?? nil, waypointShortName: node.user?.shortName ?? nil, color: Color(UIColor(hex: UInt32(node.num))))
+						CompassView(waypointLocation: latestPosition?.nodeCoordinate ?? nil, waypointLongName: node.user?.displayLongName, waypointShortName: node.user?.displayShortName, color: Color(UIColor(hex: UInt32(node.num))))
 							}
+					.sheet(isPresented: $showingDisplayNameSheet) {
+						EditNodeDisplayNameView(node: node)
+					}
+					.onReceive(NotificationCenter.default.publisher(for: NodeDisplayNameStore.didChangeNotification)) { _ in
+						displayNameRefresh += 1
+					}
```
[the remark this answers is not in the document]
Comment (garthvh) [reply Meshtastic/Views/Nodes/Helpers/NodeDetail.swift]: Fixed in 0abede4d — `NodeDisplayNameStore.didChangeNotification` now carries the changed node’s `num` as `object` (mirroring the existing `.nodeLogAvailabilityDidChange` pattern already in this file), and `NodeDetail` scopes its `onReceive` to `node.num` before bumping `displayNameRefresh`.

File: Meshtastic/Views/Nodes/Helpers/NodeListItem.swift [hunk 4]
```diff
@@ -60,9 +60,9 @@ struct NodeListItem: View {
 
 	private func accessibilityDescription(batteryLevel: Int32?, cachedLocationData: (nodeLocation: CLLocation, myLocation: CLLocation)?, status: String?) -> String {
 		var desc = ""
-		if let shortName = node.user?.shortName {
+		if let shortName = node.user?.displayShortName {
 			desc = shortName.formatNodeNameForVoiceOver()
-		} else if let longName = node.user?.longName {
+		} else if let longName = node.user?.displayLongName {
 			desc = longName
 		} else {
```
[the remark this answers is not in the document]
Comment (garthvh) [reply Meshtastic/Views/Nodes/Helpers/NodeListItem.swift:69]: Good catch — fixed in 0abede4d. Now branches on the raw `shortName`/`longName` presence (not the display-name variants, which are never nil/empty) and speaks the display-name-aware value inside each branch.

File: Meshtastic/Views/Nodes/NodeList.swift [hunk 5]
```diff
@@ -385,6 +393,11 @@ private struct FilteredNodeList: View {
 	) -> some View {
 		if let connectedNode {
 			FavoriteNodeButton(node: node)
+			Button {
+				nodeForDisplayNameEdit = node
+			} label: {
+				Label("Set display name", systemImage: "pencil.circle")
+			}
```
[the remark this answers is not in the document]
Comment (garthvh) [reply Meshtastic/Views/Nodes/NodeList.swift]: Agreed, fixed in 0abede4d — hoisted "Set display name" out of `if let connectedNode`. It's local-only and never touches the mesh, so it should be available regardless of connection state.
